README.md
# XMC Server Script
Python script for easier Minecraft server management (works properly on linux).

To run this you need:
 - Python 3.x
 - Minecraft server (default, Spigot or CraftBukkit)

## Features
- Backing up worlds
- Server stop (with timer)
- Server start
- Server restart (with timer)
- Server announcement through /say or /broadcast
- World backup restore system
- Easy update for Spigot/CraftBukkit using BuildTools
- Overall configurable

## Contributing
If you want to submit a pull request and contribute to the project, make sure to provide information as to what your PR is suppose to fix or improve.

## License
License can be found [here](https://github.com/XxMoNsTeR/XMC-Server-Script/blob/master/LICENSE.md).

## Contact
- On [Minecraft forums](https://www.minecraftforum.net/members/XxMoNsTeR)
- On [Spigot forums](https://www.spigotmc.org/members/xxmonster.359507/)

changelog.txt
--==CHANGELOG for XMC Server Script==--

--v1.6 (2020-03-19):
    -Added ability to customize start, restart, backup, restoring backup and update messages
    -Added ability to change datetime format in backup names
    -Sort backups into year and date folder structure
    -Server restarts now wait for the terminal screen to be removed before starting the server again
    -Fixed a bug which caused server to not start

--v1.5 (2017-06-26):
    -Made -b, -s, -r and -a so they can be ran without adding arguments
    -Changed format on how to save the 'worlds' and 'announce_messages'
    -Added a new check server command which tells you whether the server
     is running or not and it shows the last 20 lines of the server screen
     if it is running
    -Fixed a path bug
    -Script now automatically does a backup before updating/restoring a backup
    -Improved update input

--v1.4 (2017-04-29):
    -Added multiworld supported backups

--v1.3 (2017-03-28):
    -Added update functionality

--v1.2 (2017-03-24):
    -Added restore backup functionality

--v1.1 (2017-01-15):
    -The script won't do a backup if the server's storage device is full
    -Added output logs in the script

--v1.0 (2016-12-30):
    -Script can now detect if server is running
    -Added config file

--Pre-v1.0:
    November 2015:
        -Added basic backup, restart and announce functionality
    28 December 2016 and 29 December 2016:
        -Improved code and added start and stop functionality
        -Added version and help for the script


config.py
#Variables that are used in all files are saved in here

import configparser
import json
import utility
import sys


class ConfigException(Exception):
    def __init__(self, message):
        self._message = message

    @property
    def message(self):
        return self._message


class ConfigDefaults:
    worlds = ["world"]
    screen = "minecraft_server_xmc"
    start_server_command = "java -Xmx1024M -Xms1024M -jar minecraft_server.jar nogui"
    use_broadcast = False
    multiworld_support = False
    messages = ["message1", "message2"]
    server_tool = "craftbukkit"
    server_jar_name = "minecraft_server.jar"
    backup_folder = "backups"
    sort_backups = True

    l_datetime_format = "%Y-%m-%d_%H-%M"
    l_backup = "&6Backing up worlds"
    l_backup_as = "&6Backing up worlds as a '{0}' backup"
    l_saving_world = "&6Saving the world"
    l_backup_finished = "&6Finished the backup!"
    l_restart = "&7Sever is restarting"
    l_stopping = "&7Stopping server in {0} second{1}"
    l_restoring_backup = "&6Restoring backup"
    l_downloading_updates = "&6Downloading server updates"
    l_downloading_development_version = "&6Downloading development version"
    l_downloading_version_with_revision = "&6Downloading version with revision '{0}'"
    l_updating = "&6Updating server"

    config_file = "xmc/xmc_config.cfg"


class Config:
    def __init__(self, config_file=ConfigDefaults.config_file):
        try:
            f = open(config_file, "r")

            c = configparser.RawConfigParser()
            c.read_file(f)

            self.worlds = c.get('data', 'worlds', fallback=ConfigDefaults.worlds)
            if str(type(self.worlds)) == "<class 'str'>":
                # Convert the string to a list
                self.worlds = json.loads(self.worlds)
            self.screen = c.get('data', 'screen', fallback=ConfigDefaults.screen)
            self.start_server_command = c.get('data', 'start_server_command', fallback=ConfigDefaults.start_server_command)
            self.use_broadcast = c.getboolean('data', 'use_broadcast', fallback=ConfigDefaults.use_broadcast)
            self.multiworld_support = c.getboolean('data', 'multiworld_support', fallback=ConfigDefaults.multiworld_support)
            self.messages = c.get('data', 'announce_messages', fallback=ConfigDefaults.messages)
            if str(type(self.messages)) == "<class 'str'>":
                self.messages = json.loads(self.messages)
            self.server_tool = c.get('data', 'server_tool', fallback=ConfigDefaults.server_tool)
            self.server_jar_name = c.get('data', 'server_jar_name', fallback=ConfigDefaults.server_jar_name)
            self.backup_folder = c.get('data', 'backup_folder', fallback=ConfigDefaults.backup_folder)
            self.sort_backups = c.get('data', 'sort_backups', fallback=ConfigDefaults.sort_backups)

            self.l_datetime_format = c.get('localization', 'datetime_format', fallback=ConfigDefaults.l_datetime_format)
            self.l_backup = c.get('localization', 'backup', fallback=ConfigDefaults.l_backup)
            self.l_backup_as = c.get('localization', 'backup_as', fallback=ConfigDefaults.l_backup_as)
            self.l_saving_world = c.get('localization', 'saving_world', fallback=ConfigDefaults.l_saving_world)
            self.l_backup_finished = c.get('localization', 'backup_finished', fallback=ConfigDefaults.l_backup_finished)
            self.l_restart = c.get('localization', 'restart', fallback=ConfigDefaults.l_restart)
            self.l_stopping = c.get('localization', 'stopping', fallback=ConfigDefaults.l_stopping)
            self.l_restoring_backup = c.get('localization', 'restoring_backup', fallback=ConfigDefaults.l_restoring_backup)
            self.l_downloading_updates = c.get('localization', 'downloading_updates', fallback=ConfigDefaults.l_downloading_updates)
            self.l_downloading_development_version = c.get('localization', 'downloading_development_version',
                                                           fallback=ConfigDefaults.l_downloading_development_version)
            self.l_downloading_version_with_revision = c.get('localization', 'downloading_version_with_revision',
                                                             fallback=ConfigDefaults.l_downloading_version_with_revision)
            self.l_updating = c.get('localization', 'updating', fallback=ConfigDefaults.l_updating)

            f.close()
        except FileNotFoundError:
            utility.xmc_print("No config file found", True)
            self.generate_config(config_file)
        except ResourceWarning:
            f.close()
        except:
            raise ConfigException("An error has occured while loading the config file")

    def generate_config(self, config_file):
        utility.xmc_print("Generating a new config file")

        try:
            f = open(config_file, 'w')

            f.seek(0)
            f.truncate(0)

            f.writelines(
                ["[data]\n",
                 "worlds = {0}\n".format(utility.replace_char(str(ConfigDefaults.worlds), "'", "\"")),
                 "screen = {0}\n".format(ConfigDefaults.screen),
                 "start_server_command = {0}\n".format(ConfigDefaults.start_server_command),
                 "use_broadcast = {0}\n".format(str(ConfigDefaults.use_broadcast)),
                 "multiworld_support = {0}\n".format(str(ConfigDefaults.multiworld_support)),
                 "announce_messages = {0}\n".format(utility.replace_char(str(ConfigDefaults.messages), "'", "\"")),
                 "server_tool = {0}\n".format(ConfigDefaults.server_tool),
                 "server_jar_name = {0}\n".format(ConfigDefaults.server_jar_name),
                 "backup_folders = {0}\n".format(ConfigDefaults.backup_folder),
                 "sort_backups = {0}\n\n".format(ConfigDefaults.sort_backups),
                 "#Available options for 'server_tool' are 'craftbukkit' and 'spigot'\n",
                 "#Please use consistent names for your server jar file, otherwise the update script might not work properly\n\n",
                 "[localization]\n",
                 "datetime_format = {0}\n".format(ConfigDefaults.l_datetime_format),
                 "backup = {0}\n".format(ConfigDefaults.l_backup),
                 "backup_as = {0}\n".format(ConfigDefaults.l_backup_as),
                 "#{0} - Backup name\n",
                 "saving_world = {0}\n".format(ConfigDefaults.l_saving_world),
                 "backup_finished = {0}\n".format(ConfigDefaults.l_backup_finished),
                 "restart = {0}\n".format(ConfigDefaults.l_restart),
                 "stopping = {0}\n".format(ConfigDefaults.l_stopping),
                 "{0} - Number of seconds left until stopping the server\n",
                 "{1} - Plural for seconds, if there is more than 1 second it will be 's'\n",
                 "restoring_backup = {0}\n".format(ConfigDefaults.l_restoring_backup),
                 "downloading_updates = {0}\n".format(ConfigDefaults.l_downloading_updates),
                 "downloading_development_version = {0}\n".format(ConfigDefaults.l_downloading_development_version),
                 "downloading_version_with_revision = {0}\n".format(ConfigDefaults.l_downloading_version_with_revision),
                 "{0} - The specified revision to download given by the user through special arguments\n",
                 "updating = {0}\n".format(ConfigDefaults.l_updating)]
            )
            f.close()
            utility.xmc_print("New config file has been generated")
            self.__init__()
        except:
            raise ConfigException("Failed to generate new config file")


console.py
import utility, config


class Console:
    def __init__(self, screen, use_broadcast):
        self.__screen = screen
        self.__use_broadcast = use_broadcast

    def send_command(self, comm):
        """
        Executes a command in the console.
        """
        utility.run_command("screen -x {0} -X stuff \"{1} $(printf '\r')\"".format(self.__screen, comm))

    def say(self, message):
        self.send_command("say {0}".format(message))

    def broadcast(self, message):
        self.send_command("broadcast {0}".format(message))

    def announce(self, message):
        if self.__use_broadcast:
            self.broadcast(message)
        else:
            self.say(message)

    def save_all(self):
        self.send_command("save-all")

    def stop(self):
        self.send_command("stop")


utility.py
import os
import time
from subprocess import check_output
from subprocess import getoutput

prefix = "XMC:"
error_prefix = "XMC ERR:"
initial_time_stamp = time.strftime("%Y-%m-%d_%H:%M:%S")


def run_command(command):
    output = getoutput(command)
    if output != "":
        print(output)
    add_to_log(output)


def xmc_print(message, is_error=False):
    message = "{0} {1}".format(error_prefix if is_error else prefix, message)
    print(message)
    add_to_log(message)
    return message


def add_to_log(message):
    time_stamp = time.strftime("%Y-%m-%d_%H:%M:%S")
    if not os.path.exists("xmc/logs"):
        os.system("mkdir xmc/logs")
    if message != "":
        f = open("xmc/logs/xmc-{0}.log".format(initial_time_stamp), 'a+')
        if f.tell() == 0:
            f.write("====================================XMC LOG START===========================\n")
        f.write("[{0}]: {1}\n".format(time_stamp, message))
        f.close()


def is_screen_running(name):
    var = check_output(["screen -ls; true"], shell=True)
    return bytes(".{0}\t(".format(name), 'UTF-8') in var


def replace_char(str_, replaced_char, replace_with):
    for letter in str_:
        if letter == replaced_char:
            str_ = str_.replace(replaced_char, replace_with)
    return str_


server_control.py
import os
import random
import sys
import time
import utility

from config import Config
from console import Console
from subprocess import getoutput


class ServerController:
    def __init__(self):
        self.config = Config()
        self.__console = Console(self.config.screen, self.config.use_broadcast)
        self.screen_running = utility.is_screen_running(self.config.screen)

    def backup(self, special_name="normal", include_announce=True):
        try:
            utility.xmc_print("Backing up worlds")
            if include_announce:
                if special_name == "normal":
                    self.__console.announce(self.config.l_backup)
                else:
                    self.__console.announce(self.config.l_backup_as.format(special_name))
                self.__console.announce(self.config.l_saving_world)
            self.__console.save_all()
            if not os.path.exists(self.config.backup_folder):
                utility.run_command("mkdir -p {0}".format(self.config.backup_folder))
            path_with_special_name = "{0}/{1}".format(self.config.backup_folder, special_name)
            if not os.path.exists(path_with_special_name):
                utility.run_command("mkdir {0}".format(path_with_special_name))

            folder_str = path_with_special_name + "/"
            time_str = time.strftime(self.config.l_datetime_format)
            backup_name = "backup-{0}.tar".format(time_str)
            backup_extension = ".gz"

            if self.config.sort_backups:
                year = time.strftime("%Y")
                month = time.strftime("%m")
                path_with_year = "{0}/{1}/{2}".format(self.config.backup_folder, special_name, year)
                if not os.path.exists(path_with_year):
                    utility.run_command("mkdir {0}".format(path_with_year))
                path_with_year_and_month = "{0}/{1}".format(path_with_year, month)
                if not os.path.exists(path_with_year_and_month):
                    utility.run_command("mkdir {0}".format(path_with_year_and_month))

                folder_str = path_with_year_and_month + "/"

            if self.config.multiworld_support:
                utility.run_command("tar -cvf {0}{1} \"{2}\"".format(folder_str, backup_name, self.config.worlds[0]))
                if len(self.config.worlds) > 1:
                    for x in range(1, len(self.config.worlds)):
                        utility.run_command("tar -uvf {0}{1} \"{2}\"".format(folder_str, backup_name, self.config.worlds[x]))
                time.sleep(0.75)
                utility.run_command("gzip {0}{1}".format(folder_str, backup_name))
            else:
                utility.run_command("tar -zcvf {0}{1}{2} \"{3}\"".format(folder_str, backup_name, backup_extension, self.config.worlds[0]))

            if include_announce:
                self.__console.announce(self.config.l_backup_finished)
            utility.xmc_print("Successfully backed up worlds!")
            utility.xmc_print("The backup restore path is: '{0}{1}{2}'".format(folder_str, backup_name, backup_extension))
        except IOError as e:
            utility.xmc_print("An unexpected I/O error has occured!", True)
            utility.xmc_print(e, True)
            utility.xmc_print("Aborting backup!")
            sys.exit()

    def stop(self, time_to_stop):
        time_left = time_to_stop
        self.__console.announce(self.config.l_stopping.format(str(time_to_stop), '' if time_to_stop == 1 else 's'))
        for x in range(time_to_stop):
            time.sleep(1)
            time_left -= 1

            show_timer = False

            if time_left == 0:
                break

            if time_left <= 5:
                show_timer = True
            elif time_left == time_to_stop / 2 or time_left == time_to_stop / 4:
                show_timer = True

            if show_timer:
                self.__console.announce(self.config.l_stopping.format(str(time_left), '' if time_left == 1 else 's'))

        self.__console.save_all()
        self.__console.stop()

        utility.xmc_print("Server was stopped")

    def restart(self, time_to_restart):
        self.__console.announce(self.config.l_restart)
        self.stop(time_to_restart)
        while utility.is_screen_running(self.config.screen):
            time.sleep(0.01)
        self.start()

    def start(self):
        utility.run_command("screen -L -Logfile xmc-screen.log -dmS {0} {1}".format(self.config.screen, self.config.start_server_command))
        self.screen_running = utility.is_screen_running(self.config.screen)
        if self.screen_running:
            utility.xmc_print("Server was started")
        else:
            utility.xmc_print("Failed to start server", True)

    def announce(self, message):
        if message == "":
            msg = self.config.messages[random.randint(0, len(self.config.messages) - 1)]
            self.__console.announce("&d{0}".format(msg))
        else:
            msg = message
            self.__console.announce("&d{0}".format(msg))
        utility.xmc_print("Announced message: '{0}'".format(msg))

    def restore(self, path):
        if path[-7:] != ".tar.gz":
            path += ".tar.gz"

        if path[:8] != "backups/":
            path = "backups/" + path

        if os.path.exists(path):
            utility.xmc_print("A backup is being restored")
            self.backup("pre-restore", False)

            if self.screen_running:
                self.__console.announce(self.config.l_restoring_backup)
                time.sleep(2)
                self.__console.stop()

            archived_files = getoutput("tar -tf {0} --exclude '*/*'".format(path)).split("\n")

            for file in archived_files:
                if os.path.exists(file):
                    utility.run_command("rm -vr {0}".format(file[:-2]))

            if self.config.multiworld_support:
                utility.run_command("tar -zxvf {0}".format(path))
            else:
                archived_name_different = False
                if archived_files[0] != self.config.worlds[0]:
                    archived_name_different = True

                utility.run_command("tar -zxvf {0}".format(path))

                if archived_name_different:
                    utility.run_command("mv -vT \"{0}\" \"{1}\"".format(archived_files[0], self.config.worlds[0]))

            time.sleep(1)
            self.start()
        else:
            utility.xmc_print("Backup file doesn't exist", True)

    def update(self):
        utility.xmc_print("Please note that the updates use BuildTools.jar and only contains CraftBukkit and Spigot")
        utility.xmc_print("Make sure you have set the right server tool in xmc_config.cfg")
        utility.xmc_print("Server tool found in the config file: {0}".format(self.config.server_tool))
        utility.xmc_print("Are you sure you want to continue? (y/n)")
        response = input()
        utility.xmc_print("Response: " + response)
        if response == "y":
            utility.xmc_print("Special arguments?")
            special_args = input()
            utility.xmc_print("Response: {0}".format(special_args))
            utility.xmc_print("Starting update process...")
            self.__console.announce(self.config.l_downloading_updates)
            if not os.path.exists("BuildTools/BuildTools.jar"):
                download_link = "\"https://hub.spigotmc.org/jenkins/job/BuildTools/lastSuccessfulBuild/artifact/target/BuildTools.jar\""
                utility.run_command("wget {0} -O BuildTools.jar".format(download_link))
                utility.run_command("mkdir BuildTools")
                utility.run_command("mv BuildTools.jar BuildTools/BuildTools.jar")
            if special_args == "--dev":
                self.__console.announce(self.config.l_downloading_development_version)
            elif special_args[:5] == "--rev":
                more_args = special_args.split()
                self.__console.announce(self.config.l_downloading_version_with_revision.format(more_args[1]))

            if special_args == "":
                utility.run_command("java -jar BuildTools/BuildTools.jar")
            else:
                utility.run_command("java -jar BuildTools/BuildTools.jar {0}".format(special_args))

            self.backup("pre-update", False)
            self.__console.announce(self.config.l_updating)
            time.sleep(2)
            self.__console.save_all()
            self.__console.stop()
            time.sleep(1.5)
            utility.run_command("rm {0}".format(self.config.server_jar_name))

            filepath = ""
            jar = getoutput("ls *.jar").split("\n")

            if self.config.server_tool == "craftbukkit":
                for j in jar:
                    if j[:11] == "craftbukkit":
                        filepath = j
                        break
            elif self.config.server_tool == "spigot":
                for j in jar:
                    if j[:6] == "spigot":
                        filepath = j
                        break

            if filepath != "":
                utility.run_command("mv -f {0} {1}".format(filepath, self.config.server_jar_name))
            else:
                utility.xmc_print("Unable to find the server jar and rename it!", True)

            self.start()
        else:
            sys.exit()


start.py
#!/usr/bin/env python3

import getopt
import sys
import os
import time
import utility
import config

os.chdir(sys.path[0])
os.chdir("../")

from server_control import ServerController
from subprocess import getoutput

VERSION = "v1.6-dev7"
RELEASE_DATE = "2020-03-19"
AUTHOR = "XxMoNsTeR"


def needs_config(opt):
    return opt in ("-a", "--announce", "-b", "--backup", "-R", "--restore", "-r",
                   "--restart", "-S", "--start", "-s", "--stop", "-U", "--update",
                   "-c", "--check_config", "-e", "--check_server")


def usage():
    print("--------------------------------------------------------------")
    print("Usage:                                                        ")
    print("       -a, --announce [message]                               ")
    print("           Announce a message or type nothing to make         ")
    print("           it announce a random message from config           ")
    print("       -b, --backup [backup_name]                             ")
    print("           Backup the world file                              ")
    print("       -c, --check_config                                     ")
    print("           Checks if data is correct in the config file       ")
    print("       -e --check_server                                      ")
    print("           Shows the last 20 lines of the console output if   ")
    print("           the server is running                              ")
    print("       -g, --generate_config                                  ")
    print("           Generate a new config file (overwrites old one)    ")
    print("       -h, --help                                             ")
    print("           Shows this usage message                           ")
    print("       -R, --restore {backup_path}                            ")
    print("           Restores a specific backup                         ")
    print("       -r, --restart [restart_time]                           ")
    print("           Restart the server                                 ")
    print("       -S, --start                                            ")
    print("           Start the server                                   ")
    print("       -s, --stop [stop_time]                                 ")
    print("           Stop the server                                    ")
    print("       -U, --update                                           ")
    print("           Updates the server                                 ")
    print("       -v, --version                                          ")
    print("           Shows the version of the script                    ")
    print("--------------------------------------------------------------")
    sys.exit()


def main(argv):
    server_controller = ServerController()

    try:
        opts, args = getopt.getopt(argv, "hSUegcvbR:rsa", ["help", "start", "update", "check_server", "generate_config", "check_config", "version", "backup", "restore=", "restart", "stop", "announce"])
    except getopt.GetoptError:
        utility.xmc_print("Invalid argument", True)
        utility.xmc_print("Please use -h or --help for usage")
        sys.exit()
    for opt, arg in opts:
        try:
            screen_running = server_controller.screen_running
            args = ' '.join(args)
            if opt in ("-b", "--backup"):
                if screen_running:
                    if args == '':
                        server_controller.backup()
                    else:
                        server_controller.backup(args)
                else:
                    utility.xmc_print("Server is not started", True)
            elif opt in ("-S", "--start"):
                if not screen_running:
                    server_controller.start()
                else:
                    utility.xmc_print("Server is already started", True)
            elif opt in ("-r", "--restart"):
                if screen_running:
                    server_controller.restart(0 if args == '' else int(args))
                else:
                    server_controller.start()
            elif opt in ("-s", "--stop"):
                if screen_running:
                    server_controller.stop(0 if args == '' else int(args))
                else:
                    utility.xmc_print("Server is already stopped", True)
            elif opt in ("-a", "--announce"):
                if screen_running:
                    server_controller.announce(args)
                else:
                    utility.xmc_print("Server is not started", True)
            elif opt in ("-c", "--check_config"):
                print("Worlds: ")
                for world in server_controller.config.worlds:
                    print("   {0}".format(world))
                print("Screen: {0}".format(server_controller.config.screen))
                print("Start Server Command: {0}".format(server_controller.config.start_server_command))
                print("Use Broadcast: {0}".format(str(server_controller.config.use_broadcast)))
                print("Multiworld Support: {0}".format(str(server_controller.config.multiworld_support)))
                print("Messages: ")
                for message in server_controller.config.messages:
                    print("   {0}".format(message))
                print("Server Tool: {0}".format(server_controller.config.server_tool))
                print("Server Jar Name: {0}".format(server_controller.config.server_jar_name))
                print("Sort Backups: {0}".format(str(server_controller.config.sort_backups)))
                print("Date-Time Format: {0}".format(time.strftime(server_controller.config.datetime_format)))
            elif opt in ("-R", "--restore"):
                server_controller.restore(arg)
            elif opt in ("-U", "--update"):
                server_controller.update()
            elif opt in ("-e", "--check_server"):
                if screen_running:
                    utility.xmc_print("Server is running")
                    utility.xmc_print("Here are the last 20 lines of the Console Output:")
                    utility.xmc_print(getoutput("tail -20 xmc-screen.log"))
                else:
                    utility.xmc_print("Server is not running")
        except config.ConfigException as e:
            utility.xmc_print(e)

        if opt in ("-h", "--help"):
            usage()
        elif opt in ("-v", "--version"):
            print("XMC Server Script {0}".format(VERSION))
            print("Release Date: {0}".format(RELEASE_DATE))
            print("Author: {0}".format(AUTHOR))
        elif opt in ("-g", "--generate_config"):
            server_controller.config.generate_config(config.ConfigDefaults.config_file)


if __name__ == "__main__":
   main(sys.argv[1:])


